### PR TITLE
Add wait_for_all_acked support

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Related to ros2/rmw#295

### CODE REVIEW

File: rclpy/rclpy/publisher.py [hunk 1]
```diff
@@ -100,3 +101,21 @@ def assert_liveliness(self) -> None:
...
+        data were acknowledged, but not waiting.
+        If the timeout is greater than 0 then this function will return after that period of time
+        has elapsed or all published message data were acknowledged.
+
+        :param timeout: the duration to wait for all published message data were acknowledged
+        :returns: true if all published message data were acknowledged before timeout, otherwise
+            false.
+        """
+        with self.handle:
+            return self.__publisher.wait_for_all_acked(timeout._duration_handle)
```
Comment (fujitatomoya) [inline rclpy/rclpy/publisher.py:121]: it should use `try&except` statement? Publisher::wait_for_all_acked would throw RCLError exception.

File: rclpy/test/test_publisher.py [hunk 2]
```diff
@@ -75,6 +76,26 @@ def do_test_topic_name(cls, test_topics, node):
             assert publisher.topic_name == target_topic
             publisher.destroy()
 
+    @classmethod
+    def do_test_wait_for_all_acked(cls, test_topic, node):
+        pub = node.create_publisher(BasicTypes, test_topic, 1)
```
Comment (fujitatomoya) [inline rclpy/test/test_publisher.py:81]: QoS Reliable setting would be nice to check too.

File: rclpy/rclpy/publisher.py (hunk 1 above)
Comment (Barry-Xu-2018) [reply rclpy/rclpy/publisher.py:121]: After check other codes, I find it also doesn't deal with RCLError.   
e.g.

https://github.com/ros2/rclpy/blob/714f69ee64ee8ea5dbc5f3067211b7886a4e7226/rclpy/rclpy/publisher.py#L59-L73
Not catch exception

The underlying code may throw exception

https://github.com/ros2/rclpy/blob/714f69ee64ee8ea5dbc5f3067211b7886a4e7226/rclpy/src/rclpy/publisher.cpp#L121-L133

**Comment by Barry-Xu-2018** [comment]
Friendly ping @fujitatomoya.  Please look at my comments and I supplement tests.

**Comment by fujitatomoya** [comment]
CC: @ivanpauno @wjwwood friendly ping.

File: rclpy/test/test_publisher.py [hunk 3]
```diff
@@ -103,6 +124,27 @@ def test_topic_name_remapping(self):
...
+        pub_qos.reliability = rclpy.qos.QoSReliabilityPolicy.BEST_EFFORT
+        sub_qos.reliability = rclpy.qos.QoSReliabilityPolicy.BEST_EFFORT
+        TestPublisher.do_test_wait_for_all_acked(TEST_TOPIC, self.node, pub_qos, sub_qos)
+
+        # Publisher is reliable and subscription is best effort
+        pub_qos.reliability = rclpy.qos.QoSReliabilityPolicy.RELIABLE
+        TestPublisher.do_test_wait_for_all_acked(TEST_TOPIC, self.node, pub_qos, sub_qos)
+
+        # Publisher is best effort and subscription is reliable
+        # Not work.
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/test/test_publisher.py]: ac0a979

File: rclpy/test/test_publisher.py [hunk 4]
```diff
@@ -103,6 +124,27 @@ def test_topic_name_remapping(self):
...
+        TestPublisher.do_test_wait_for_all_acked(TEST_TOPIC, self.node, pub_qos, sub_qos)
+
+        # Publisher and subscription are all best effort
+        pub_qos.reliability = rclpy.qos.QoSReliabilityPolicy.BEST_EFFORT
+        sub_qos.reliability = rclpy.qos.QoSReliabilityPolicy.BEST_EFFORT
+        TestPublisher.do_test_wait_for_all_acked(TEST_TOPIC, self.node, pub_qos, sub_qos)
+
+        # Publisher is reliable and subscription is best effort
+        pub_qos.reliability = rclpy.qos.QoSReliabilityPolicy.RELIABLE
+        TestPublisher.do_test_wait_for_all_acked(TEST_TOPIC, self.node, pub_qos, sub_qos)
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/test/test_publisher.py]: ac0a979

File: rclpy/test/test_publisher.py [hunk 5]
```diff
@@ -75,6 +76,26 @@ def do_test_topic_name(cls, test_topics, node):
...
+        assert pub.get_subscription_count() == 1
+        basic_types_msg = BasicTypes()
+        try:
+            pub.publish(basic_types_msg)
+        except Exception:
+            raise AssertionError('Publish data failed')
+
+        try:
+            assert pub.wait_for_all_acked() is True
+        except Exception:
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/test/test_publisher.py]: ac0a979

File: rclpy/test/test_publisher.py [hunk 6]
```diff
@@ -75,6 +76,26 @@ def do_test_topic_name(cls, test_topics, node):
...
+        time.sleep(1)
+        assert pub.get_subscription_count() == 1
+        basic_types_msg = BasicTypes()
+        try:
+            pub.publish(basic_types_msg)
+        except Exception:
+            raise AssertionError('Publish data failed')
+
+        try:
+            assert pub.wait_for_all_acked() is True
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/test/test_publisher.py]: ac0a979

File: rclpy/test/test_publisher.py [hunk 7]
```diff
@@ -75,6 +76,26 @@ def do_test_topic_name(cls, test_topics, node):
             assert publisher.topic_name == target_topic
             publisher.destroy()
 
+    @classmethod
+    def do_test_wait_for_all_acked(cls, test_topic, node, pub_qos, sub_qos):
+        pub = node.create_publisher(BasicTypes, test_topic, pub_qos)
+        sub = node.create_subscription(BasicTypes, test_topic, lambda msg: print(msg), sub_qos)
+        time.sleep(1)
```
[the remark this answers is not in the document]
Comment (sloretz) [reply rclpy/test/test_publisher.py]: ac0a979

**Review by sloretz — APPROVED** [review]
LGTM, since it's been a while I made some changes pushed it to your branch directly

I would like it to be reviewed by once more by someone who hasn't edited this PR

File: rclpy/src/rclpy/publisher.cpp [hunk 8]
```diff
@@ -143,6 +143,18 @@ Publisher::publish_raw(std::string msg)
...
+bool
+Publisher::wait_for_all_acked(rcl_duration_t pytimeout)
+{
+  rcl_ret_t ret = rcl_publisher_wait_for_all_acked(rcl_publisher_.get(), pytimeout.nanoseconds);
+  if (RCL_RET_OK == ret) {
+    return true;
+  } else if (RCL_RET_TIMEOUT == ret) {
+    return false;
+  }
+  throw RCLError("Failed to wait for all acknowledgements");
```
Comment (ivanpauno) [inline rclpy/src/rclpy/publisher.cpp:155]: shouldn't this use the rcl error message?

**Review by ivanpauno — APPROVED** [review]

File: rclpy/src/rclpy/publisher.cpp (hunk 8 above)
Comment (ivanpauno) [reply rclpy/src/rclpy/publisher.cpp:155]: Maybe using [append_rcl_error](https://github.com/ros2/rclpy/blob/a1a874742e76850c5a9e04c8392104e1e2e78121/rclpy/src/rclpy/exceptions.cpp#L34) (?).

Comment (sloretz) [reply rclpy/src/rclpy/publisher.cpp:155]: RCLError calls `append_rcl_error()` internally: https://github.com/ros2/rclpy/blob/a6aac563af6fa9825d8ea686ede6ad6df74d8a16/rclpy/src/rclpy/exceptions.cpp#L56

Comment (ivanpauno) [reply rclpy/src/rclpy/publisher.cpp:155]: ah, didn't see that.
Thanks!

**Review by ivanpauno — APPROVED** [review]

**Comment by Barry-Xu-2018** [comment]
rcl package isn't updated (ros2/rcl#913 isn't included). So CI failed. 
```
14:41:44 /tmp/ws/src/rclpy/rclpy/src/rclpy/publisher.cpp: In member function ‘bool rclpy::Publisher::wait_for_all_acked(rcl_duration_t)’:
14:41:44 /tmp/ws/src/rclpy/rclpy/src/rclpy/publisher.cpp:149:19: error: ‘rcl_publisher_wait_for_all_acked’ was not declared in this scope
14:41:44   149 |   rcl_ret_t ret = rcl_publisher_wait_for_all_acked(rcl_publisher_.get(), pytimeout.nanoseconds);
14:41:44       |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
14:41:44 make[2]: *** [CMakeFiles/_rclpy_pybind11.dir/build.make:297: CMakeFiles/_rclpy_pybind11.dir/src/rclpy/publisher.cpp.o] Error 1
14:41:44 make[1]: *** [CMakeFiles/Makefile2:193: CMakeFiles/_rclpy_pybind11.dir/all] Error 2
14:41:44 make: *** [Makefile:141: all] Error 2
```

**Comment by Barry-Xu-2018** [comment]
@ros-pull-request-builder retest this please

**Comment by ivanpauno** [comment]
@Barry-Xu-2018 @fujitatomoya  can you run ci?
This is ready to go

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15202/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9922/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12882/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15383/)

**Review by ivanpauno — APPROVED** [review]

File: rclpy/test/test_publisher.py [hunk 9]
```diff
@@ -103,6 +105,29 @@ def test_topic_name_remapping(self):
         ]
         TestPublisher.do_test_topic_name(test_topics, self.node)
 
+    def test_wait_for_all_acked(self):
```
Comment (jacobperron) [inline rclpy/test/test_publisher.py:108]: This test is failing with Connext https://github.com/ros2/rclpy/issues/824

**Pull request merged**